pupil_src/shared_modules/observable.py

```diff
@@ -0,0 +1,320 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2018 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+
+import functools
+import inspect
+import weakref
+
+
+class Observable:
+    """
+    An object that allows to add observers to any of its methods. See
+    the docstring for add_observer for more information.
+    """
+
+    def add_observer(self, method_name, observer):
+        """
+        Adds an observer to a method. An observer is an arbitrary callable. Every time
+        the method is invoked, the observer will be called.
+
+        The observer is called with the same arguments as the method, so make sure
+        that the observer can handle these arguments!
+
+        You cannot add observers to static and class methods.
+
+        It's possible to add multiple observers to the same method.
+
+        Observers are called AFTER the method completed. Even if the method raises an
+        exception, observers are still called.
+        Observers should NOT raise exceptions. If they do, exceptions are forwarded
+        to the caller of the method.
+
+        You only add the observer for this object. The observer is not triggered for
+        the same method of a different object.
```

Comment:
My feeling is when you just state a method and a callback, then people could think you are patching it for everywhere, so I just added it to make it clearer. You are right that this will be clear to most people, but I think this short sentence is ok here.